src/auth.py

```diff
@@ -0,0 +1,134 @@
+import base64
+import hashlib
+import requests
+import random
+import string
+import datetime
+import urllib
+
+from requests_toolbelt.multipart.encoder import MultipartEncoder
+
+from bb2 import Bb2
+
+BB2_AUTH_URL = "{}/v{}/o/authorize"
+BB2_TOKEN_URL = "{}/v{}/o/token/"
+
+
+class AuthRequest:
+
+    def __init__(self, bb: Bb2):
+        self.bb = bb
+        self.auth_base_url = BB2_AUTH_URL.format(bb.get_config()['base_url'], bb.get_config()['version'])
+        self.auth_token_url = BB2_TOKEN_URL.format(bb.get_config()['base_url'], bb.get_config()['version'])
+        self.auth_data = self._generate_authdata()
+        self.auth_url = self._generate_authorize_url()
+        self.auth_token = None
+
+    def get_authorize_url(self):
+        return self.auth_url
+
+    def authorize_callback(self, code, state):
+        if code is None:
+            raise ValueError("Authorization code missing.")
+
+        if state is None:
+            raise ValueError("Callback parameter 'state' missing.")
+
+        if state != self.auth_data['state']:
+            raise ValueError("Provided callback state does not match.")
+
+        self.auth_token = AuthorizationToken(self._get_access_token(code))
+
+        return self.auth_token
+
+    def access_token_expired(self):
+        return self.auth_token.access_token_expired()
+
+    def refresh_access_token(self):
+
+        if self.auth_token.refresh_token is None:
+            raise ValueError("Refresh token not available when calling refresh_access_token().")
```

Comment:
I feel like this error message might be well served if there was a recommended fix. What should the user do if that is the case? Is there a way to programmatically enforce this?

Reply:
Not sure what action on SDK side could be more helpful...
Since mostly use case for refresh_access_token() is called during fhir query when cached access token expires, SDK could retry the underlying post on 5xx...

Reply:
@JFU-GIT @nbragdon Should this check be removed, similar to the Node SDK?  Instead it might be better for the error response to come directly from the BB2 server side?